### PR TITLE
Initial controller/vmm decoupling

### PR DESCRIPTION
**Pull request by alexandruag**
Initial decoupling implementation. Still working on adapting tests.

Signed-off-by: Alexandru Agache <aagch@amazon.com>

### CODE REVIEW

File: src/vmm/src/controller.rs [hunk 1]
```diff
@@ -0,0 +1,999 @@
...
+use vmm_config::vsock::{VsockDeviceConfig, VsockError};
+
+/// Enables pre-boot setup, instantiation and real time configuration of a Firecracker VMM.
+pub struct VmmController {
+    device_configs: DeviceConfigs,
+    epoll_context: EpollContext,
+    kernel_config: Option<KernelConfig>,
+    vm_config: VmConfig,
+    shared_info: Arc<RwLock<InstanceInfo>>,
+    vmm: Option<Vmm>,
```
Comment (sandreim) [inline src/vmm/src/controller.rs]: I do not like this option here. The question that comes up is about the identity of chicken and egg.
The way I see things, the Controller is built on top of a VMM while the VMM si built by the builder.

I would like to see:
```rust
let builder = VMMBuilder::new();

// configure builder 
builder = builder.with_vcpus(vcpu_config).with_memory(mem_config);

// conditionally build stuff based on user configuration
if let Some(block_configs) = self.vmconfig.get_block_configuration() {
     builder = builder.with_block_devs(block_configs);
}

// Or alternatively, interface the builder with the API server so we do all of the above
// on API call and we can just call .run() on the builder in StartMicrovm.
// This came up in the original discussions about using a builder pattern approach
// for VM configuration.

VmmController::new(builder.run() ...)
```

Comment (acatangiu) [reply src/vmm/src/controller.rs]: I had some comments on the same line. This is a step in the right direction but I also strongly believe we should go for finer separation of concerns and have a _Builder_ that **just** builds each resources including the `Vmm` and a _Controller_ which does the runtime Vmm button-pushing.

Having said that, we could also do those changes incrementally and not block this PR seeing how we're waaaaay behind the decoupling timeline. In its current form this PR unblocks other pending work.

Comment (sandreim) [reply src/vmm/src/controller.rs]: I was not saying that I expect the Builder to be implemented in this PR, just some more polish that is in the direction of that code snippet.

Comment (alexandruag) [reply src/vmm/src/controller.rs]: Thanks for the comments! These initial steps take away as much code that's Firecracker specific as possible from `vmm/lib.rs`, and move it somewhere else with the end goal of having a generic and reusable Vmm at some point in the future. Moreover, it becomes easier to work on things in isolation along the way.  

Granted, the use of the `VmmBuilder` name here was unfortunate; in this context it was just something that allowed devices to be attached before the vmm gets to run, and then later starts and returns the vmm. It also singles out a set of parameters the generic vmm requires from the outside; in this case the contents of `VmmBuilderConfig` (we can likely get rid of the kernel command line). 

I think the point where we entirely replace `EpollContext` with polly is when we'll have a much better idea about how things interact overall. The builder will actually be something on the Firecracker side that, well, builds things while the generic side will use a different nomenclature and flow (especially if we also start using the approach where devices register themselves). I can add further refinements here if they help moving forward, but the main point was to pick some low hanging fruits in terms of disentangling the old vmm code.

Comment (sandreim) [reply src/vmm/src/controller.rs]: After all I think it's better to go with what we have right here as @acatangiu pointed out. We can incrementally do whatever it makes sense as we reach that point.

**Pull request merged**